sdk/ml/azure-ai-ml/azure/ai/ml/operations/_component_operations.py

```diff
@@ -290,7 +291,11 @@ def create_or_update(
             component = _refine_component(component)
         if version is not None:
             component.version = version
-        if not component.version and component._auto_increment_version:
+        # In non-registry scenario, if component does not have version, no need to get next version here.
+        # As Component property version has setter that updates `_auto_increment_version` in-place, then
+        # a component will get a version after its creation, and it will always use this version in its
+        # future creation operations, which breaks version auto increment mechanism.
+        if self._registry_name and not component.version and component._auto_increment_version:
```

Comment:
@brynn-code  Please take a look at this PR before we merge this.

Reply:
ACK